core/src/main/java/com/google/errorprone/bugpatterns/StringCaseLocaleUsage.java

```diff
@@ -0,0 +1,73 @@
+package com.google.errorprone.bugpatterns;
+
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.BugPattern.StandardTags.FRAGILE_CODE;
+import static com.google.errorprone.matchers.method.MethodMatchers.instanceMethod;
+import static com.sun.tools.javac.parser.Tokens.TokenKind.RPAREN;
+
+import com.google.common.collect.Streams;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.fixes.Fix;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.util.ASTHelpers;
+import com.google.errorprone.util.ErrorProneTokens;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.util.Position;
+
+/**
+ * A {@link BugChecker} that flags calls to {@link String#toLowerCase()} and {@link
+ * String#toUpperCase()}, as these methods implicitly rely on the environment's default locale.
+ */
+@BugPattern(
+    summary = "Specify a `Locale` when calling `String#to{Lower,Upper}Case`",
+    severity = WARNING,
+    tags = FRAGILE_CODE)
+public final class StringCaseLocaleUsage extends BugChecker implements MethodInvocationTreeMatcher {
+  private static final Matcher<ExpressionTree> DEFAULT_LOCALE_CASE_CONVERSION =
+      instanceMethod()
+          .onExactClass(String.class.getName())
+          .namedAnyOf("toLowerCase", "toUpperCase")
+          .withNoParameters();
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    if (!DEFAULT_LOCALE_CASE_CONVERSION.matches(tree, state)) {
+      return Description.NO_MATCH;
+    }
+
+    int closingParenPosition = getClosingParenPosition(tree, state);
+    if (closingParenPosition == Position.NOPOS) {
+      return describeMatch(tree);
+    }
+
+    return buildDescription(tree)
+        .addFix(suggestLocale(closingParenPosition, "Locale.ROOT"))
+        .addFix(suggestLocale(closingParenPosition, "Locale.getDefault()"))
+        .build();
+  }
+
+  private static Fix suggestLocale(int insertPosition, String locale) {
+    return SuggestedFix.builder()
+        .addImport("java.util.Locale")
+        .replace(insertPosition, insertPosition, locale)
+        .build();
+  }
+
+  private static int getClosingParenPosition(MethodInvocationTree tree, VisitorState state) {
```

Comment:
Just curious, did you see examples like `.toUpperCase /* Comment with parens: (). */()` in practice?

It's nice to handle this correctly, but I would normally have just done something like `replace(getEndPosition(tree.getMethodSelect()), getEndPosition(tree), "(Locale.ROOT)")`

Reply:
Hmm, no we didn't encounter a lot of problems, we created this after our initial implementation (that was based on a `String#replaceFirst`) which wasn't correct enough. 

It's indeed a bit more complex than what you are proposing. We can do 3 things:
(1) Extract this functionality to a utility class and re-use this setup, which might make it worth that it is slightly more complex.  (2) Keep it as-is.
(3) Use the method you are proposing.

Let me know what you prefer :smile:.

Reply:
I'd suggest creating a helper in `SuggestedFixes`, but I'm not sure how general it is without a way to choose where the argument is inserted, and I don't have a great ideas for the API. Maybe something like `addArgumentToMethod` with a parameter for the insertion index?

How about `(2)` for now, and if you feel like it you can add a `TODO` to consider promoting it to `SuggestedFixes` later :)

Reply:
Deal, let's go for `(2)`, I'll add a `TODO`.